modules/evm/src/lib.rs

```diff
@@ -1033,27 +1033,58 @@ impl<T: Config> Pallet<T> {
 				T::NetworkContractSource::get()
 			};
 
-			let code_size = code.len() as u32;
-			let code_hash = code_hash(&code.as_slice());
+			let old_code_info = Self::code_infos(&contract_info.code_hash).ok_or(Error::<T>::ContractNotFound)?;
+
+			let bounded_code: BoundedVec<u8, T::MaxCodeSize> =
+				code.try_into().map_err(|_| Error::<T>::ContractExceedsMaxCodeSize)?;
+			let code_hash = code_hash(&bounded_code.as_slice());
+			let code_size = bounded_code.len() as u32;
+			// The code_hash of the same contract is definitely different.
+			// The `contract_info.code_hash` hashed by on_contract_initialization which constructored.
+			// Still check it here.
 			if code_hash == contract_info.code_hash {
 				return Ok(());
 			}
 
-			ensure!(
-				code_size <= T::MaxCodeSize::get(),
-				Error::<T>::ContractExceedsMaxCodeSize
-			);
+			let storage_size_chainged: i32 =
+				code_size.saturating_add(T::NewContractExtraBytes::get()) as i32 - old_code_info.code_size as i32;
+			let mut handler = StorageMeterHandlerImpl::<T>::new(source);
+			if storage_size_chainged.is_positive() {
+				handler.reserve_storage(storage_size_chainged as u32)?;
+				handler.charge_storage(&contract, storage_size_chainged as u32, 0)?;
+			} else {
+				handler.charge_storage(&contract, 0, storage_size_chainged as u32)?;
```

Comment:
```suggestion
				handler.charge_storage(&contract, 0, -storage_size_chainged as u32)?;
```

Also need more tests?